### PR TITLE
Remove implicit forking in ZIO#onDone and ZIO#onDoneCause to align with onX methods

### LINKED ISSUE #9191: Consider deprecating `ZIO#onDone`
I was debugging a nasty bug at $WORK, and to my surprise I found out that `ZIO#onDone` forks the effect and "discards" the fiber, running the success/error callbacks on exit (see [here](https://github.com/zio/zio/blob/cf4a2fc691247a5a60efcd48de9efb44945ae496/core/shared/src/main/scala/zio/ZIO.scala#L1046-L1060))

I think this behaviour is quite dangerous because its behaviour is _very_ different than other `onX` methods (e.g., `onExit`). As a user (or in my case, when I reviewed the PR that caused the bug), I won't know that the method forks the effect and doesn't await on its completion unless I take a look at the source code.

In order to guard against incorrect usage of this method, we should consider deprecating it (`onDoneCause` as well). If we find out that users do use this method in the correct way and need it, we can introduce it under a more accurate name (`forkDaemonExit` or something)

### PR DESCRIPTION
**Pull request by asr2003**
Updated onDone and onDoneCause to execute callbacks synchronously on the main fiber to align behavior with other onX methods removing implicit forking( (i.e. NOT fork).

Closes #9191 
/claim #9191 

### Important Note about Changes to `onDone` and `onDoneCause` Behavior

In this release, we have updated the behavior of `onDone` and `onDoneCause` to address concerns raised in [Issue #9191](https://github.com/zio/zio/issues/9191). Previously, these methods operated asynchronously by forking their callback effects, which could lead to subtle bugs and race conditions. This behavior was unique among ZIO's `onX` methods and diverged from typical expectations.

#### Key Behavior Change: Synchronous Execution

Starting in this release, `onDone` and `onDoneCause` now execute success and failure callbacks synchronously within the calling fiber. This update aligns them with other `onX` methods (e.g., `onExit`), ensuring that the effect completes fully, including any specified callbacks, before moving on. This change provides greater predictability and reliability by removing the previously implicit asynchronous behavior.

#### Summary and Impact

This enhancement addresses the risk identified in [Issue #9191](https://github.com/zio/zio/issues/9191), where the unexpected forking could cause surprising, hard-to-trace bugs. As a result, users may notice different callback behaviors following this upgrade. Where necessary, alternative patterns (such as explicit forking) can be employed to replicate prior behavior. We encourage users to review any code that relies on `onDone` or `onDoneCause` and verify that the updated synchronous behavior aligns with expected outcomes.

### CODE REVIEW

**Comment by jdegoes** [comment]
@asr2003 These need tests to guarantee synchronous behavior for each.

**Comment by asr2003** [comment]
@jdegoes Sure! I will add it

**Review by kyri-petrou — COMMENTED** [review]
@asr2003 in addition to John's comment, can you please do the following:

1. Add scaladoc to both `onDone` and `onDoneCause` methods
2. Write a paragraph explaining the behaviour change of `onDone` which we will use in release notes. This should be similar to the "Important note about ZPool changes" found in the [v2.1.10 notes](https://github.com/zio/zio/releases/tag/v2.1.10)

**Comment by kyri-petrou** [comment]
One last comment, I think the implementation should use `onExit` so that we're consistent across `onX` implementations

**Comment by asr2003** [comment]
@kyri-petrou @jdegoes Done with the changes!

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -1043,21 +1043,38 @@ sealed trait ZIO[-R, +E, +A]
...
     error: Cause[E] => ZIO[R1, Nothing, Any],
     success: A => ZIO[R1, Nothing, Any]
   )(implicit trace: Trace): ZIO[R1, Nothing, Unit] =
-    ZIO.uninterruptibleMask { restore =>
-      restore(self).foldCauseZIO(e => restore(error(e)), s => restore(success(s))).forkDaemon.unit
-    }
+    onExit {
+      case Exit.Success(value) => success(value).unit
+      case Exit.Failure(cause) => error(cause).unit
+    }.catchAll(_ => ZIO.unit).as(())
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/ZIO.scala]: The `catchAll(_ => ZIO.unit).as(())` part is not needed
```suggestion
    onExit {
      case Exit.Success(value) => success(value).unit
      case Exit.Failure(cause) => error(cause).unit
    }
```

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 2]
```diff
@@ -1043,21 +1043,38 @@ sealed trait ZIO[-R, +E, +A]
...
+    onExit {
+      case Exit.Success(value) => success(value).unit
+      case Exit.Failure(cause) =>
+        cause.failureOrCause
+          .fold(
+            error,
+            _ => ZIO.unit
+          )
+          .unit
+    }.catchAll(_ => ZIO.unit).as(())
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/ZIO.scala]: Reuse `onDoneCause`:

```suggestion
    onDoneCause(
      _.failureOrCause.fold(error, Exit.failCause(_)),
      success
    )
```

**Comment by asr2003** [comment]
@kyri-petrou There has been type error if we use `OnExit` directly on `OnDoneCause`

```
error] /home/runner/work/zio/zio/core/shared/src/main/scala/zio/ZIO.scala:1074:12: type mismatch;
[error]  found   : zio.ZIO[R1,E,A]
[error]  required: zio.ZIO[R1,Nothing,Unit]
[error]     onExit {
[error]            ^
```

So I have updated for the same to resolve this type mismatch

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 3]
```diff
@@ -1043,21 +1043,32 @@ sealed trait ZIO[-R, +E, +A]
...
     error: Cause[E] => ZIO[R1, Nothing, Any],
     success: A => ZIO[R1, Nothing, Any]
   )(implicit trace: Trace): ZIO[R1, Nothing, Unit] =
-    ZIO.uninterruptibleMask { restore =>
-      restore(self).foldCauseZIO(e => restore(error(e)), s => restore(success(s))).forkDaemon.unit
-    }
+    onExit {
+      case Exit.Success(value) => success(value).unit
+      case Exit.Failure(cause) => error(cause).unit
+    }.catchAll(_ => ZIO.unit).as(())
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/ZIO.scala]: You have too many units here, which will decrease performance. Try:

```scala
    onExit {
      case Exit.Success(value) => success(value)
      case Exit.Failure(cause) => error(cause)
    }.catchAll(_ => ZIO.unit).as(())
```

If `onExit` requires unit callbacks, just change it to require `ZIO[..., Any]` instead, which should be backward compatible.

Comment (asr2003) [reply core/shared/src/main/scala/zio/ZIO.scala]: Done

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 4]
```diff
@@ -1043,21 +1043,32 @@ sealed trait ZIO[-R, +E, +A]
...
     error: Cause[E] => ZIO[R1, Nothing, Any],
     success: A => ZIO[R1, Nothing, Any]
   )(implicit trace: Trace): ZIO[R1, Nothing, Unit] =
-    ZIO.uninterruptibleMask { restore =>
-      restore(self).foldCauseZIO(e => restore(error(e)), s => restore(success(s))).forkDaemon.unit
-    }
+    onExit {
+      case Exit.Success(value) => success(value)
+      case Exit.Failure(cause) => error(cause)
+    }.catchAll(_ => ZIO.unit).as(())
```
Comment (kyri-petrou) [inline core/shared/src/main/scala/zio/ZIO.scala]: Simpler and more performant:

```suggestion
    }.ignore
```

**Comment by asr2003** [comment]
@kyri-petrou Done with the changes!

**Review by kyri-petrou — APPROVED** [review]

**Pull request merged**